File: server/crypto.js
const crypto = require('node:crypto');

const RESOURCE = GetCurrentResourceName();
const SECRET_FILE = '.secret';

const SCRYPT_N = 16384;
const SCRYPT_R = 8;
const SCRYPT_P = 1;
const KEY_LEN = 32;
const SALT_LEN = 16;
const IV_LEN = 12;

function loadSecret() {
    const convar = GetConvar('sd_phone_secret', '');
    if (convar && convar.trim() !== '') {
        return crypto.createHash('sha256').update(convar.trim(), 'utf8').digest();
    }

    const saved = LoadResourceFile(RESOURCE, SECRET_FILE);
    if (saved && /^[0-9a-f]{64}$/i.test(saved.trim())) {
        return Buffer.from(saved.trim(), 'hex');
    }

    const generated = crypto.randomBytes(32);
    SaveResourceFile(RESOURCE, SECRET_FILE, generated.toString('hex'), -1);
    console.log(
        '^3[sd-phone:crypto]^0 no `sd_phone_secret` convar set, so a server secret was generated ' +
        `into ${RESOURCE}/${SECRET_FILE}. Back this file up: losing it makes saved Passwords ` +
        'unreadable and forces every account through a password reset.'
    );
    return generated;
}

const SECRET = loadSecret();
const PEPPER = crypto.createHmac('sha256', SECRET).update('pepper').digest();
const VAULT_KEY = crypto.createHmac('sha256', SECRET).update('vault').digest();

function peppered(plain) {
    return Buffer.concat([Buffer.from(String(plain), 'utf8'), PEPPER]);
}

function hashPassword(plain) {
    const salt = crypto.randomBytes(SALT_LEN);
    const dk = crypto.scryptSync(peppered(plain), salt, KEY_LEN, { N: SCRYPT_N, r: SCRYPT_R, p: SCRYPT_P });
    return `scrypt$${SCRYPT_N}$${SCRYPT_R}$${SCRYPT_P}$${salt.toString('base64')}$${dk.toString('base64')}`;
}

function verifyPassword(plain, stored) {
    if (typeof stored !== 'string') return false;
    const parts = stored.split('$');
    if (parts.length !== 6 || parts[0] !== 'scrypt') return false;
    try {
        const want = Buffer.from(parts[5], 'base64');
        const dk = crypto.scryptSync(peppered(plain), Buffer.from(parts[4], 'base64'), want.length, {
            N: Number(parts[1]), r: Number(parts[2]), p: Number(parts[3]),
        });
        return dk.length === want.length && crypto.timingSafeEqual(dk, want);
    } catch (err) {
        return false;
    }
}

function encrypt(plain) {
    if (typeof plain !== 'string' || plain === '') return plain;
    const iv = crypto.randomBytes(IV_LEN);
    const cipher = crypto.createCipheriv('aes-256-gcm', VAULT_KEY, iv);
    const body = Buffer.concat([cipher.update(plain, 'utf8'), cipher.final()]);
    return `v1$${iv.toString('base64')}$${cipher.getAuthTag().toString('base64')}$${body.toString('base64')}`;
}

function decrypt(blob) {
    if (typeof blob !== 'string') return null;
    const parts = blob.split('$');
    if (parts.length !== 4 || parts[0] !== 'v1') return null;
    try {
        const decipher = crypto.createDecipheriv('aes-256-gcm', VAULT_KEY, Buffer.from(parts[1], 'base64'));
        decipher.setAuthTag(Buffer.from(parts[2], 'base64'));
        return Buffer.concat([decipher.update(Buffer.from(parts[3], 'base64')), decipher.final()]).toString('utf8');
    } catch (err) {
        return null;
    }
}

// FiveM's JS runtime injects `exports` as the registration function; plain CommonJS binds it to
// the module's exports object instead. Resolving both keeps the file loadable off-server.
const registerExport = typeof exports === 'function' ? exports : global.exports;

// lb-phone hashed with bcrypt, which Node has no built-in for. Verifying it is what lets a
// migrated player sign in with the password they already know instead of being locked out of an
// account nobody recorded the owner of. Verify only: nothing here ever writes a bcrypt hash, and
// the accounts engine rewrites the account to scrypt the first time one is accepted.
let bcrypt = null;
try {
    bcrypt = require('./server/vendor/bcryptjs.js');
} catch (err) {
    console.log(`^3[sd-phone:crypto]^0 bcrypt helper did not load (${err.message}); lb-phone passwords cannot be verified.`);
}

const BCRYPT_RE = /^\$2[abxy]\$\d{2}\$[./A-Za-z0-9]{53}$/;

function verifyBcrypt(plain, stored) {
    if (!bcrypt || typeof plain !== 'string' || typeof stored !== 'string') return false;
    if (!BCRYPT_RE.test(stored)) return false;
    try {
        return bcrypt.compareSync(plain, stored) === true;
    } catch (err) {
        return false;
    }
}

registerExport('sdCryptoReady', () => true);
registerExport('sdCryptoHashPassword', hashPassword);
registerExport('sdCryptoVerifyPassword', verifyPassword);
registerExport('sdCryptoVerifyBcrypt', verifyBcrypt);
registerExport('sdCryptoEncrypt', encrypt);
registerExport('sdCryptoDecrypt', decrypt);
registerExport('sdCryptoSha256', (s) => crypto.createHash('sha256').update(String(s), 'utf8').digest('hex'));
registerExport('sdCryptoRandomHex', (n) => crypto.randomBytes(Math.min(64, Math.max(1, Number(n) || 32))).toString('hex'));

// The relay token helpers. `server/media/tokens.lua` signs every grant through these, and
// `media-server/src/token.js` verifies what comes out, so the two implementations have to agree
// byte for byte: HMAC-SHA256 over the ASCII string `sdmr1.<base64url(claims)>`, with the raw
// signature bytes (not their hex) base64url'd as the third part.
const RELAY_KEY_RE = /^[0-9a-f]{64}$/;

function relayKey(keyHex) {
    return typeof keyHex === 'string' && RELAY_KEY_RE.test(keyHex) ? Buffer.from(keyHex, 'hex') : null;
}

function b64url(buf) {
    return buf.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

registerExport('sdCryptoHmacHex', (keyHex, msg) => {
    const key = relayKey(keyHex);
    if (!key || typeof msg !== 'string') return null;
    return crypto.createHmac('sha256', key).update(msg, 'utf8').digest('hex');
});

registerExport('sdCryptoRelayToken', (keyHex, payloadJson) => {
    const key = relayKey(keyHex);
    if (!key || typeof payloadJson !== 'string' || payloadJson === '') return null;
    const part = b64url(Buffer.from(payloadJson, 'utf8'));
    const mac = crypto.createHmac('sha256', key).update(`sdmr1.${part}`, 'utf8').digest();
    return `sdmr1.${part}.${b64url(mac)}`;
});

File: server/crypto.lua
---@type table Crypto module; the table returned at end of file.
local crypto = {}

---@type string This resource's name, used to reach the Node helper's exports.
local RESOURCE = GetCurrentResourceName()

---@type boolean|nil Cached result of the helper probe; nil until first use.
local ready

---Calls one of the Node helper's exports, swallowing a missing runtime.
---@param name string export name
---@param ... any arguments
---@return boolean okCall, any result
local function call(name, ...)
    return pcall(function(...) return exports[RESOURCE][name](exports[RESOURCE], ...) end, ...)
end

---True when the Node crypto helper answered. Probed once, then cached; a miss is announced so a
---server on a runtime without it is not silently left on the legacy hash.
---@return boolean available
function crypto.available()
    if ready == nil then
        local okCall, res = call('sdCryptoReady')
        ready = okCall and res == true
        if not ready then
            print('^1[sd-phone:crypto]^0 Node crypto helper did not load. Passwords stay on the legacy hash and saved Passwords stay unencrypted.')
        end
    end
    return ready
end

---Hashes a password with scrypt. Nil when the helper is unavailable.
---@param plain string plaintext password
---@return string|nil hash
function crypto.hashPassword(plain)
    if not crypto.available() then return nil end
    local okCall, res = call('sdCryptoHashPassword', plain)
    return (okCall and type(res) == 'string') and res or nil
end

---Checks a plaintext password against a scrypt hash.
---@param plain string plaintext password
---@param stored string stored scrypt hash
---@return boolean verified
function crypto.verifyPassword(plain, stored)
    if not crypto.available() then return false end
    local okCall, res = call('sdCryptoVerifyPassword', plain, stored)
    return okCall and res == true
end

---Encrypts a vault secret. Nil when the helper is unavailable.
---@param plain string
---@return string|nil blob
function crypto.encrypt(plain)
    if not crypto.available() then return nil end
    local okCall, res = call('sdCryptoEncrypt', plain)
    return (okCall and type(res) == 'string') and res or nil
end

---Decrypts a vault secret. Nil when the helper is unavailable or the blob is not ours.
---@param blob string
---@return string|nil plain
function crypto.decrypt(blob)
    if not crypto.available() then return nil end
    local okCall, res = call('sdCryptoDecrypt', blob)
    return (okCall and type(res) == 'string') and res or nil
end

---Hashes an arbitrary string with SHA-256, lowercase hex. Nil when the helper is unavailable, so
---a caller that needs a verifiable commitment can degrade openly instead of publishing a fake one.
---@param s string value to hash
---@return string|nil hex 64 lowercase hex characters
function crypto.sha256(s)
    if not crypto.available() then return nil end
    local okCall, res = call('sdCryptoSha256', s)
    return (okCall and type(res) == 'string') and res or nil
end

---Cryptographically strong random bytes as lowercase hex. Nil when the helper is unavailable.
---@param bytes integer byte count, clamped to 1..64 by the helper
---@return string|nil hex 2 * bytes lowercase hex characters
function crypto.randomHex(bytes)
    if not crypto.available() then return nil end
    local okCall, res = call('sdCryptoRandomHex', bytes)
    return (okCall and type(res) == 'string') and res or nil
end

return crypto
